lib/rules/func-name-matching.js

```diff
@@ -0,0 +1,149 @@
+        ]
+    },
+
+    create(context) {
+
+        const includeModuleExports = context.options[0] && context.options[0].includeCommonJSModuleExports;
+        const ecmaVersion = context.parserOptions && context.parserOptions.ecmaVersion ? context.parserOptions.ecmaVersion : 5;
+
+        /**
+         * Reports
+         * @param {ASTNode} node The node to report
+         * @param {string} name The variable or property name
+         * @param {string} funcName The function name
+         * @param {boolean} isProp True if the reported node is a property assignment
+         * @returns {void}
+         */
+        function report(node, name, funcName, isProp) {
+            context.report({
+                node,
+                message: isProp ? "Function name `{{funcName}}` should match property name `{{name}}`"
+                                : "Function name `{{funcName}}` should match variable name `{{name}}`",
+                data: {
+                    name,
+                    funcName
+                }
+            });
+        }
+
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+
+        return {
+
+            VariableDeclarator(node) {
+                if (node.init.type !== "FunctionExpression") {
+                    return;
+                }
+                if (node.init.id && node.id.name !== node.init.id.name) {
+                    report(node, node.id.name, node.init.id.name, false);
+                }
+            },
+
+            AssignmentExpression(node) {
+                if (node.right.type !== "FunctionExpression" ||
+                   (node.left.computed && node.left.property.type !== "Literal") ||
+                    (!includeModuleExports && isModuleExports(node.left))
+                   ) {
+                    return;
+                }
+
+                const isProp = node.left.type === "MemberExpression" ? true : false;
+                const name = isProp ? astUtils.getStaticPropertyName(node.left) : node.left.name;
+
+                if (node.right.id && isIdentifier(name) && name !== node.right.id.name) {
+                    report(node, name, node.right.id.name, isProp);
+                }
+            },
+
+            Property(node) {
+                if (node.computed || node.value.type !== "FunctionExpression" || !node.value.id) {
+                    return;
+                }
+                if (node.key.type === "Identifier" && node.key.name !== node.value.id.name) {
+                    report(node, node.key.name, node.value.id.name, true);
+                } else if (node.key.type === "Literal" &&
+                           isIdentifier(node.key.value, ecmaVersion) &&
+                           node.key.value !== node.value.id.name
+                          ) {
+                    report(node, node.key.value, node.value.id.name, true);
```

Comment:
Could you add tests for this `if` block?

- valid
    - `({["foo"]: function foo() {}})`
    - `({["❤"]: function foo() {}})`
    - `({[foo]: function bar() {}})`
- invalid
    - `({["foo"]: function bar() {}})`